### DISCUSSION TITLE
Cross Compile with QT for raspberry pi

### DISCUSSION THREAD

**Discussion by ben80** [wang-bin/mdk-sdk#249] [General]
Hello, i just came across mdk-sdk and it might be what i am searching for a long time.
Currently i am displaying videos on a raspberry pi with hardware aceleration and would like to replace the "video tools" that i am using.
I cross compiled qt5 for raspberry pi and use POT ([Link](https://bugfreeblog.duckdns.org/build-procedure-for-pot)) to show the video. Unfortunately this is not very stable. Could someone answer my following questions :

- Can i cross-compile mdk-sdk with QT for raspberry pi?
- Can i use QTs qml code to play videos?
- Will it show videos  hw accelerated (no sound needed)?
- does it work with qt5 and/or qt6?

Thanks a lot
Ben

**Comment by wang-bin** [comment]
download [mdk-sdk-linux.tar.xz](https://sourceforge.net/projects/mdk-sdk/files/nightly/mdk-sdk-linux.tar.xz/download) and extract it, include `$MDK_SDK_DIR/mdk-sdk/lib/cmake/FindMDK.cmake` or `$MDK_SDK_DIR/mdk-sdk/mdk.pri` in your cmake or qmake project, then you should be able to build your project, cross or native compile. 

qml examples can be found [here](https://github.com/wang-bin/mdk-examples/tree/master/Qt).

hw accelerated decoding and rendering are supported, delete libffmpeg.so.* in mdk-sdk to use system ffmpeg to support h264, hevc hardware decoder, use opengl es2/3 to support 0-copy rendering.

both qt5 and 6 are supportd.

**Reply by ben80** [reply]
Thanks a lot, sounds great. I will look into it.
Do I use the same files on my host-pc and also on the raspberry pi? Sorry if this is a dummy question. I am always confused when cross compiling.
And do i need to cross compile ffmpeg as well or basically find the right sources/binaries for my host-pc and also for the raspberry pi?

**Reply by wang-bin** [reply]
for linux, you can use the same package mdk-sdk-linux.tar.xz, it contains 3 cpu archtectures: x86_64(amd64), arm64, armhf. cmake and qmake should find out the right one. when deploying, you have to copy at least libmdk.so.0 in mdk/lib/${cpu}. `libffmpeg.so.*` is also there, you can also use system installed ffmpeg(version >=4.4) by deleting `libffmpeg.so.*`. there is a libc++.so.1, you can use the one in system or installed by package manager.

**Reply by ben80** [reply]
I am finally there to integrate mdk-sdk into my program.
Unfortunately i need some basic help.

RPI4
QT6.9
qmultimedia should use mdk-sdk to get a better performance when playing movie files (hardware accelerated)

What i did:
1.  Copied mdk-sdk-linux.tar.xz to my host-linux-computer (host) and also to the raspberry pi (PI)
`   /opt/mdk-sdk in both cases`
2. Added the following line to my <projectname>.pro file (host)
`   INCLUDEPATH += /opt/mdk-sdk/mdk.pri`
3. Crosscompiled my application and deployed it to PI
4. Using` ldd <myProgram>` (PI) doesn't show any link to mdk-sdk

So i guess i am doing something (a lot ?) wrong

Would be nice to get some help
Thank you

**Reply by wang-bin** [reply]
include($$MDK_SDK_DIR/mdk.pri)

**Reply by ben80** [reply]
I changed it, i can see the mdk-sdk in the qtcreator project tree. I did a clean rebuild and i can see references to mdk-sdk within the compile output (no errors). How can i check if it actually worked?

**Reply by wang-bin** [reply]
play a video. for rpi, additional settings are required to get best performance, see https://github.com/wang-bin/mdk-sdk?tab=readme-ov-file#recommended-settings

**Reply by ben80** [reply]
i created a very simple prog with `mdk::Player player;`
The output is as below. It means, it is working correctly in principle, right?

My target is to play video using qml (qt6). Currently i am using the `Video `qml-element.
It doesn't seem the qml-element is included in _mdk-sdk-linux.tar.xz_.
Which steps do i need next to play a video with qml?



```
0.33.1 (git 8da6cdd) - Multimedia Development Kit. Copyright (c) 2016-2025 WangBin(QtAV author) <wbsecg1 at gmail.com>
Build for: debian/12 4KB ARMv8 AArch64; glibc2.24/2.36 libc++210000; Clang21.0.0 (++20250702083737+272576543246-1~exp1~20250702203908.1008); 15:51:48 Jul  5 2025. Raspberry Pi 4 Model B Rev 1.5
Player started. State: 0
Current state: 0
Current state: 0
Current state: 1
Current state: 1
..
Current state: 1
Current state: 1
Current state: 1
Player stopped.
```

**Reply by wang-bin** [reply]
you need a qtmultimedia plugin if use qt's `Video`, i write one: https://github.com/wang-bin/qtmultimedia-plugins-mdk

but qt6 is not supported yet. so you have to use qml or qmlrhi example: https://github.com/wang-bin/mdk-examples/tree/master/Qt

**Reply by ben80** [reply]
What is the difference between qml, qmlrhi and qmlrhi0 examples?

When do you think the qtmuldimedia plugin will be ready for qt6?

Thanks

**Reply by wang-bin** [reply]
qml is based on opengl framebuffer, supports old qt versions. qmlrhi is based on rhi, qt >= 5.14 is required. qmlrhi0 is not maintaned. qmlrhi is recommended

**Reply by ben80** [reply]
Thanks, does "qml" (openly framebuffer) also work with qt6 or only qmlrhi? The qmlrhi example contains several "todos".

**Reply by wang-bin** [reply]
qml supports qt6. there are always some todos

**Reply by ben80** [reply]
I am making progress. I added some of your "qml-example" code to my app and it displays a video.
I added this line to the play() function:
`internal_player->setDecoders(MediaType::Video, {"V4L2M2M", "FFmpeg:hwcontext=drm", "FFmpeg", "dav1d"});`

I attached a log with your logging. Would you mind having a look at it?
Anything i should work on?

Thanks so far for your help.
[mdk-sdk_log1.txt](https://github.com/user-attachments/files/22084503/mdk-sdk_log1.txt)

**Comment by wang-bin** [comment]
> [FFmpeg:h264_v4l2m2m] Using device /dev/video10
0x5582a61d60 ***buffering progress 56%***
[FFmpeg:h264_v4l2m2m] driver 'bcm2835-codec' on card 'bcm2835-codec-decode' in mplane mode
[FFmpeg:h264_v4l2m2m] requesting formats: output=H264/none capture=YU12/yuv420p
0x5582a61d60 ***buffering progress 76%***
02.09.2025  00:11:33.194   RX: SET_LOG_LEVEL 1 
0x5582a61d60 ***buffering progress 96%***
0x5582a61d60 ***buffering progress 100%***
default 0x5582a61d60 FrameReader::update MediaStatus 0X14=>0X124
++++++++++++BUFFERING END++++++++++++
AVCodec.Video[h264_v4l2m2m.] decoder: 1 Single threads
unsupported avcodec options:  hwaccel=v4l2m2m
h264 1280x720  yuv420p
ERROR! failed to setup video decoder

failed to use v4l2m2m decoder. can you try to delete libffmpeg.so* and try again?

**Reply by ben80** [reply]
I moved lib/arm64/libffmpeg.so.8 and got the following output.
[mdk-sdk_log2.txt](https://github.com/user-attachments/files/22101119/mdk-sdk_log2.txt)

**Reply by wang-bin** [reply]
this looks good, it's using hardware decoder
